### ISSUE TITLE
CubDB crash with :error, :emfile

### ISSUE THREAD

**Issue by meyercm** [lucaong/cubdb#30]
My app updates the same 3 keys in CubDB at 1Hz.

Every 4.5 hours of normal running (exactly 16,435 seconds in several instances), I get a crash in CubDB, which supervision restarts:

```
14:32:12.527 [error] Process #PID<0.2058.0> terminating
** (MatchError) no match of right hand side value: {:error, :emfile}
    (cubdb 1.0.0-rc.5) lib/cubdb/clean_up.ex:42: CubDB.CleanUp.handle_cast/2
    (stdlib 3.11.1) gen_server.erl:637: :gen_server.try_dispatch/4
    (stdlib 3.11.1) gen_server.erl:711: :gen_server.handle_msg/6
    (stdlib 3.11.1) proc_lib.erl:249: :proc_lib.init_p_do_apply/3
Initial Call: CubDB.CleanUp.init/1
Ancestors: [:db, MyApp.Supervisor, #PID<0.2054.0>]
Message Queue Length: 0
Messages: []
Links: [#PID<0.2056.0>]
Dictionary: []
Trapping Exits: false
Status: :running
Heap Size: 6772
Stack Size: 27
Reductions: 30001910
Neighbours:
    #PID<0.32690.3>
        Initial Call: anonymous fn/0 in CubDB.Store.File.create/1
        Current Call: :gen_server.loop/7
        Ancestors: [:db, MyApp.Supervisor, #PID<0.2054.0>]
        Message Queue Length: 0
        Links: [#PID<0.2056.0>]
        Trapping Exits: false
        Status: :waiting
        Heap Size: 1598
        Stack Size: 11
        Reductions: 1500
        Current Stacktrace:
            (stdlib 3.11.1) gen_server.erl:394: :gen_server.loop/7
            (stdlib 3.11.1) proc_lib.erl:249: :proc_lib.init_p_do_apply/3
    #PID<0.32695.3>
        Initial Call: anonymous fn/0 in CubDB.Store.File.create/1
        Current Call: :gen_server.loop/7
        Ancestors: [:db, MyApp.Supervisor, #PID<0.2054.0>]
        Message Queue Length: 0
        Links: [#PID<0.2056.0>]
        Trapping Exits: false
        Status: :waiting
        Heap Size: 987
        Stack Size: 11
        Reductions: 52624
        Current Stacktrace:
            (stdlib 3.11.1) gen_server.erl:394: :gen_server.loop/7
            (stdlib 3.11.1) proc_lib.erl:249: :proc_lib.init_p_do_apply/3
    #PID<0.32733.0>
        Initial Call: anonymous fn/0 in CubDB.Store.File.create/1
        Current Call: :gen_server.loop/7
        Ancestors: [:db, MyApp.Supervisor, #PID<0.2054.0>]
        Message Queue Length: 0
        Links: [#PID<0.2056.0>]
        Trapping Exits: false
        Status: :waiting
        Heap Size: 610
        Stack Size: 11
        Reductions: 52413
        Current Stacktrace:
            (stdlib 3.11.1) gen_server.erl:394: :gen_server.loop/7
            (stdlib 3.11.1) proc_lib.erl:249: :proc_lib.init_p_do_apply/3
        ...
```

**The Neighbors list continues and is eventually truncated by the logging, but includes many copies of the 3 listed above**

---

When I look in /proc/<pid>/fd, I see MANY broken symlinks to deleted CubDB files. The last few look like this: (2 valid, 1 referencing an old DB version - there are hundreds more of the broken ones.

lrwx------. 1 <username> <username> 64 Feb  8 07:28 986 -> /home/<username>/<project_name>/_build/prod/rel/<project_name>/dev_db/1B06.cub
lrwx------. 1 <username> <username> 64 Feb  8 07:28 987 -> /home/<username>/<project_name>/_build/prod/rel/<project_name>/dev_db/1B06.cub
lrwx------. 1 <username> <username> 64 Feb  8 07:28 99 -> /home/<username>/<project_name>/_build/prod/rel/<project_name>/dev_db/194B.cub (deleted)      

Further digging reveals that 2 new fds are created every 30-35 seconds, which corresponds with the timing I see above for exhausting the system limit for FDs.

My best guess, quickly looking at the code is that the `Actor` that holds state for `CubDB.Store.File` isn't getting cleaned up after his version of the DB is out of date.

I'm going to continue investigating and will update with what I find.

**Comment by meyercm** [comment]
Looks like this doesn't occur in the latest release (RC7). It was present in RC5.

**Comment by meyercm** [comment]
Apparently I celebrated too early - this is still occurring in rc.7 and rc.8.  

Digging into the codebase, I can see where `Store.File.close` would actually close the file handle, but that method doesn't appear to be called outside of `BTree.stop`, which only appears in the `terminate` callback of the top-level CubDB process.  

It looks like the root cause is that `CleanUp.remove_older_files` just calls File.rm without calling `Store.File.close`.

I'm happy to write a PR, but there might need to be some discussion about how to find the actors associated with the old files (if they exist) so that they can release their FDs before removing the file. My natural reaction would be to have a map of filename => Store.File somewhere accessible to both `Store.File.create` and `CleanUp.remove_older_files`.  If you have thoughts on how you'd like to see it implemented, feel free to share, I'm going to fist focus on writing a test that reproduces the issue.

**Comment by lucaong** [comment]
Hi @meyercm ,
Thanks a lot for your investigation! It looks you’re on to something. Later today I will follow your lead and debug this.

**Comment by lucaong** [comment]
Hi @meyercm ,
Thanks a lot for your investigation! It looks you’re on to something. Later today I will follow your lead and debug this.

**Comment by meyercm** [comment]
Updates:

* I can replicate the crash by just `CubDB.compact()`ing an empty database 1024 times if I wait for the `:compaction_complete` message in between.
* It looks like there are two problems:
  1. The file is opened and never closed.  In a quick test branch, I brought in `:gproc` to keep track of all the file agents and called `close` before the `File.rm` step in cleanup.  That reduced the number of dead file descriptors by 1/2.
  2. The compacted file is never closed.  `File.rename/2` leaves the original file descriptor open, but now pointing at the new file. 

I'll push what I've got in just a bit.

**Comment by lucaong** [comment]
So, you are right, the `Store` is not closed during cleanup. There is already a mechanism in place to ensure that cleanup only happens after old files are not referenced anymore, so no need to implement it. Also, stopping a `Btree` already closes the `Store`. Basically, the problem is that at the end of the compaction/catch up process `CubDB` should stop the old `Btree`, but it doesn’t. Those `Btree` processes (and their `Store`) therefore stay there hanging.

This problem only manifest itself after many compactions (in your test 1024) without a restart of the containing application, when the leaked processes are too many. This is why the issue went undetected so far. Luckily the fix is quite trivial, I will commit it soon and create a new release.

Thanks so much for the bug report and great investigation, you discovered an important and sneaky bug and uncovered its causes, making `CubDB` better!

**Comment by meyercm** [comment]
Sorry to double tap - added a pr (#33) and came back to find this comment.  Feel free to use or discard as you like.

**Comment by lucaong** [comment]
Thank you for the PR! What do you think about this? https://github.com/lucaong/cubdb/pull/34

Basically, the `CubDB` genserver keeps a list of old `Btree`s to be stopped as soon as no reader is referencing their file anymore. It then loops through them and stops them (closing their `Store` and releasing the file descriptor), unless they were already stopped for any reason.

**Comment by lucaong** [comment]
Fixed in release `v1.0.0-rc.9`. Thanks again for the outstanding contribution!
